Add debug-only use of new AssemblyBuilder.Save in Regex.CompileToAssembly

stephentoub opened the pull request:
To aid in debugging RegexCompiler issues and to help vet the new AssemblyBuilder.Save support.

This brings back and then tweaks code that was removed in https://github.com/dotnet/runtime/pull/59734.

stephentoub commented on src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.CompileToAssembly.Tests.cs:
```diff
@@ -11,27 +13,66 @@ namespace System.Text.RegularExpressions.Tests
+
+            string assemblyPath = Path.Combine(TestDirectory, assemblyName + ".dll");
+
+            // TODO https://github.com/dotnet/runtime/issues/15704: Currently fails with Format of the executable (.exe) or library (.dll) is invalid.
+
+            //Assembly a = Assembly.LoadFrom(assemblyPath);
+            //foreach ((RegexCompilationInfo rci, string validInput, string invalidInput) in regexes)
+            //{
+            //    Regex r = (Regex)Activator.CreateInstance(a.GetType($"{rci.Namespace}.{rci.Name}"));
+            //    Assert.True(r.IsMatch(validInput));
+            //    Assert.False(r.IsMatch(invalidInput));
+            //}
```
@buyaa-n, this fails with a load exception stating "Currently fails with Format of the executable (.exe) or library (.dll) is invalid".  Is that expected at this point?  You can uncomment this and run the regex function tests to diagnose if it's not expected.

buyaa-n replied:
This should have fixed with this PR https://github.com/dotnet/runtime/pull/96362 that merged yesterday, not sure if its using the fixed version ...

stephentoub replied:
When I run with the latest System.Reflection.Emit, then I get this:
```C#
Process terminated. Assertion failed.
     at System.Reflection.Emit.ILGeneratorImpl.AdjustDepth(OpCode opcode, LabelInfo label) in D:\repos\runtime\src\libraries\System.Reflection.Emit\src\System\Reflection\Emit\ILGeneratorImpl.cs:line 444
     at System.Reflection.Emit.ILGeneratorImpl.Emit(OpCode opcode, Label label) in D:\repos\runtime\src\libraries\System.Reflection.Emit\src\System\Reflection\Emit\ILGeneratorImpl.cs:line 475
     at System.Text.RegularExpressions.RegexCompiler.BrtrueFar(Label l) in D:\repos\runtime\src\libraries\System.Text.RegularExpressions\src\System\Text\RegularExpressions\RegexCompiler.cs:line 268
     at System.Text.RegularExpressions.RegexCompiler.EmitScan(RegexOptions options, MethodInfo tryFindNextStartingPositionMethod, MethodInfo tryMatchAtCurrentPositionMethod) in D:\repos\runtime\src\libraries\System.Text.Regular
  Expressions\src\System\Text\RegularExpressions\RegexCompiler.cs:line 5403
     at System.Text.RegularExpressions.RegexAssemblyCompiler.GenerateRegexType(String pattern, RegexOptions options, String name, Boolean isPublic, RegexTree tree, RegexInterpreterCode code, TimeSpan matchTimeout) in D:\repos\r
  untime\src\libraries\System.Text.RegularExpressions\src\System\Text\RegularExpressions\RegexAssemblyCompiler.cs:line 66
     at System.Text.RegularExpressions.Regex.CompileToAssembly(RegexCompilationInfo[] regexinfos, AssemblyName assemblyname, CustomAttributeBuilder[] attributes, String resourceFile) in D:\repos\runtime\src\libraries\System.Tex
  t.RegularExpressions\src\System\Text\RegularExpressions\Regex.cs:line 256
     at System.Text.RegularExpressions.Regex.CompileToAssembly(RegexCompilationInfo[] regexinfos, AssemblyName assemblyname) in D:\repos\runtime\src\libraries\System.Text.RegularExpressions\src\System\Text\RegularExpressions\Re
  gex.cs:line 223
     at System.Text.RegularExpressions.Tests.RegexCompileToAssemblyTests.CompileToAssembly_SimpleUseInDebug() in D:\repos\runtime\src\libraries\System.Text.RegularExpressions\tests\FunctionalTests\Regex.CompileToAssembly.Tests.
  cs:line 58
```
It's tripping over this assert:
https://github.com/dotnet/runtime/blob/5a2d1519b01b2935e343b6aba2989f97ed596914/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ILGeneratorImpl.cs#L444

stephentoub replied:
And if I disable the assert, then when I try to use the assembly, I get:
```
  Unhandled exception. System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
   ---> System.InvalidProgramException: Common Language Runtime detected an invalid program.
     at Namespace1.Type1..ctor()
     at System.RuntimeType.CreateInstanceDefaultCtor(Boolean publicOnly, Boolean wrapExceptions) in D:\repos\runtime\src\coreclr\System.Private.CoreLib\src\System\RuntimeType.CoreCLR.cs:line 3914
```

buyaa-n replied:
Hm, I wonder if the code trying to run the assembly created (invoke methods, create instance), the new `AssemblyBuilder` is only for saving, after saving you can load the saved assembly with `Assembly.LoadFrom(assemblyPath);` or so and run.

stephentoub replied:
Actually, the latter looks like user error on my part, and the presence of the invalid IL was triggering the assert. We should ensure we don't have asserts that fire with invalid IL.

stephentoub commented:
@buyaa-n, the build is failing with:
```
/_/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs(190,9): Trim analysis error IL2121: System.Reflection.Emit.ModuleBuilderImpl.WriteInterfaceImplementations(TypeBuilderImpl, TypeDefinitionHandle): Unused 'UnconditionalSuppressMessageAttribute' for warning 'IL2072'. Consider removing the unused warning suppression. [D:\a\_work\1\s\src\libraries\sfx.proj]
/_/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs(182,9): Trim analysis error IL2121: System.Reflection.Emit.TypeBuilderImpl.CheckInterfaces(Type[]): Unused 'UnconditionalSuppressMessageAttribute' for warning 'IL2075'. Consider removing the unused warning suppression. [D:\a\_work\1\s\src\libraries\sfx.proj]
```
Have you seen these?

stephentoub committed:
Add debug-only use of new AssemblyBuilder.Save in Regex.CompileToAssembly

To aid in debugging RegexCompiler issues and to help vet the new AssemblyBuilder.Save support.

stephentoub commented:
@buyaa-n, I fixed the invalid code gen, but now I'm hitting this:
```
Unhandled exception. System.MissingMethodException: Method not found: 'System.ReadOnlySpan`1<Char> System.ReadOnlySpan`1.Slice(Int32)'
```
Is there something related to generics that's not yet implemented maybe?

buyaa-n commented:
> @buyaa-n, the build is failing with:
> 
> ```
> /_/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs(190,9): Trim analysis error IL2121: System.Reflection.Emit.ModuleBuilderImpl.WriteInterfaceImplementations(TypeBuilderImpl, TypeDefinitionHandle): Unused 'UnconditionalSuppressMessageAttribute' for warning 'IL2072'. Consider removing the unused warning suppression. [D:\a\_work\1\s\src\libraries\sfx.proj]
> /_/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs(182,9): Trim analysis error IL2121: System.Reflection.Emit.TypeBuilderImpl.CheckInterfaces(Type[]): Unused 'UnconditionalSuppressMessageAttribute' for warning 'IL2075'. Consider removing the unused warning suppression. [D:\a\_work\1\s\src\libraries\sfx.proj]
> ```
> 
> Have you seen these?

No, never seen those and removing those suppressions causing build failure:  
```log
...\src\System\Reflection\Emit\TypeBuilderImpl.cs(184,49): error IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Type.GetMethods(BindingFlags)'. The return value of method 'System.Collections.IEnumerator.Current.get' does not have matching annotations.
 The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

buyaa-n commented:
> Is there something related to generics that's not yet implemented maybe?

No generics should have covered completely, wonder how that happen, more logs could be helpful

stephentoub commented:
@sbomer, can you help @buyaa-n with these illinker warnings?

stephentoub commented:
I opened https://github.com/dotnet/runtime/issues/96469 for the span issue.

buyaa-n commented:
> No, never seen those warnings

Now I do see the warnings with your branch, but weirdly removing those suppressions reintroduces the IL2075 warning back (not the IL2072)
```log
...\src\System\Reflection\Emit\TypeBuilderImpl.cs(184,49): error IL2075: 'this' argument does not satisfy 'DynamicallyAcc
essedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.NonPublicMethods' in call to 'System.Type.GetMethods(BindingFlags)'. The return value of method 'System.Col
lections.IEnumerator.Current.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location
it is assigned to.
```

> I opened https://github.com/dotnet/runtime/issues/96469 for the span issue.

Thanks, I will fix it ASAP

sbomer committed:
Fix IL2121 warnings

stephentoub commented:
@sbomer investigated the linker issue and narrowed it down to https://github.com/dotnet/runtime/issues/96646. He's working on a fix.

stephentoub commented on src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs:
```diff
@@ -176,13 +176,14 @@ private void ValidateAllAbstractMethodsAreImplemented()
         {
             foreach (Type interfaceType in _interfaces)
             {
+#pragma warning disable IL2075 // Analyzer produces a different warning code than illink. The IL2065 suppression takes care of illink: https://github.com/dotnet/runtime/issues/96646
```
```suggestion
#pragma warning disable IL2075 // Analyzer produces a different warning code than illink. The IL2075 suppression takes care of illink: https://github.com/dotnet/runtime/issues/96646
```

stephentoub commented:
> @sbomer
[Fix IL2121 warnings](https://github.com/dotnet/runtime/pull/96462/commits/dad81dbe8dca4702c80ef97fab59a16789d7c2ba)

Thanks.

stephentoub commented:
@buyaa-n, can you take another look?

buyaa-n commented on src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.CompileToAssembly.Tests.cs:
```diff
@@ -33,5 +38,34 @@ public void CompileToAssembly_PNSE()
+        [ConditionalFact(nameof(IsDebug))]
+        public void CompileToAssembly_SimpleUseInDebug()
+        {
+            (RegexCompilationInfo rci, string validInput, string invalidInput)[] regexes =
```
Just curios, seems the `validInput` and `invalidInput` is not used/tested/asserted. Is the compiled assembly did not work for validation or that part is coming later?

stephentoub replied:
Yup, they're not currently used. I could remove them, but left them in as I'd like to revisit getting the test to actually run the regex. The problem right now is there are multiple aspects of the implementation that assume in-memory only, such as generating IL that accesses private types in corelib. At some point when I have some free cycles I'll look into how minimal I can make tweaks in order to be able to successfully run them; the downside to that is, with the primary purpose of this being diagnostic, changes that diverge from the code that's used by RegexOptions.Compiled makes it potentially harder to diagnose issues with it.

buyaa-n reviewed (approved):
Looks great to me, thanks for using new AssemblyBuilder implementation.
